daemon/containerd/progress.go

```diff
@@ -0,0 +1,167 @@
+package containerd
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/remotes"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+)
+
+type updateProgressFunc func(ctx context.Context, ongoing *jobs, output progress.Output, start time.Time) error
+
+func showProgress(ctx context.Context, ongoing *jobs, out progress.Output, updateFunc updateProgressFunc) func() {
+	stop := make(chan struct{})
+	ctx, cancelProgress := context.WithCancel(ctx)
+
+	var (
+		ticker = time.NewTicker(100 * time.Millisecond)
+		start  = time.Now()
+	)
+
+	for _, j := range ongoing.Jobs() {
+		id := stringid.TruncateID(j.Digest.Encoded())
+		progress.Update(out, id, "Preparing")
+	}
+
+	go func() {
+		defer func() {
+			ticker.Stop()
+			stop <- struct{}{}
+		}()
+
+		for {
+			select {
+			case <-ticker.C:
+				if !ongoing.IsResolved() {
+					continue
+				}
+				err := updateFunc(ctx, ongoing, out, start)
+				if err != nil {
+					logrus.WithError(err).Error("Updating progress failed")
+					return
+				}
+
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return func() {
+		cancelProgress()
+		<-stop
+	}
+}
+
+func pullProgress(cs content.Store, showExists bool) updateProgressFunc {
+	return func(ctx context.Context, ongoing *jobs, out progress.Output, start time.Time) error {
```

Comment:
The boolean argument is not used _for _now_ but will once I start upstreaming the [push code](https://github.com/rumpl/moby/blob/c8d/daemon/containerd/image_push.go#L87) but I agree with you, booleans as arguments are not great, I'll se to change it to something more readable

Reply:
Perhaps replace `updateProgressFunc` with `progressUpdater interface { UpdateProgress(context.Context, *jobs, progress.Output, time.Time) error }` so that `pullProgress` could be a struct type?
```go
type pullProgress struct {
        Store content.Store
        ShowExists bool
}
```

It would be just as composable: `combineProgress` could be implemented as `type combineProgress []progressUpdater`

Reply:
I was actually thinking of completely removing the `combineProgress`, containerd uses [wrapper handlers](https://github.com/containerd/containerd/blob/main/remotes/handlers.go#L274) for this kind of things so we could do the same I think